### DISCUSSION TITLE
IJob, ILogger, dependency injection confusion.

### DISCUSSION THREAD

**Discussion by s3YwCf2ZbfJG4SHAfjQMAjtsf** [quartznet/quartznet#1914] [Q&A]
I am using Quartz 3.5 with .Net 6
The Scheduler is defined just before app.RunAsync in Program.cs or later.   We don't use "AddQuartz" because the database containing the Quartz tables may be defined after app start.  Sample of SchedulerBuilder.Create below.

**Error**:
2023-02-13 10:48:01,050 [3] ERROR Quartz.Core.ErrorLogger - An error occurred instantiating job to be executed. job= 'MySite.MyJob'
Quartz.SchedulerException: Problem instantiating class 'MySite.Jobs.MyJob: Cannot instantiate type which has no empty constructor (Parameter 'MyJob')'

According to other messages and the sample code, this isn't supposed to be necessary: UseMicrosoftDependencyInjectionScopedJobFactory.   

**MyJob**
```
        private readonly ILogger<MyJob> _logger;

        public HsMergeJob(ILogger<MyJob> logger)
        {
            _logger = logger;
        }
```

Is there something I am missing in the **SchedulerBuilder.Create** syntax below?   

**Code to generate the scheduler**
```
                // and override values via builder
                _scheduler = await SchedulerBuilder.Create(properties)
                    // https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html#quartz-scheduler-instanceid
                    // Unique per Instance.   
                    .WithId(SchedulerIstanceId)
                    // https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html#quartz-scheduler-instancename
                    // Unique per Cluster "Group"
                    .WithName(SchedulerGroupName)
                    // default max concurrency is 1
                    .UseDefaultThreadPool(x => x.MaxConcurrency = 1)
                    // this is the default 
                    // .WithMisfireThreshold(TimeSpan.FromSeconds(60))
                    .UsePersistentStore(x =>
                    {
                        // force job data map values to be considered as strings
                        // prevents nasty surprises if object is accidentally serialized and then 
                        // serialization format breaks, defaults to false
                        x.UseProperties = true;
                        x.UseClustering();
                        // there are other SQL providers supported too 
                        // Another reason why the Schema User and Schema Name need to match.
                        // Oracle Connection strings do not allow you to set the Schema Name.
                        x.UseOracle(/* connection string here */);
                        // this requires Quartz.Serialization.Json NuGet package
                        x.UseJsonSerializer();
                    })
                    .BuildScheduler();

                await _scheduler.Start();
```

Thanks for any help you can provide.

Peter

**Comment by lahma** [comment]
You probably need to set `JobFactory` property of the scheduler you've built with new instance of `MicrosoftDependencyInjectionJobFactory`. That's one of the things `AddQuartz()` does behind the scenes.

`UseJobFactory` might also be available on the builder.

**Reply by s3YwCf2ZbfJG4SHAfjQMAjtsf** [reply]
Unfortunately, that didn't work.   I tried using the UseJobFactory line when creating the scheduler   I also tried configuring it via AddQuartz in 

                    .UseJobFactory<MicrosoftDependencyInjectionJobFactory>()

Now it is failing on "BuildScheduler()"

```
                _scheduler = await SchedulerBuilder.Create(properties)
                    // https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html#quartz-scheduler-instanceid
                    // Unique per Instance.   
                    .WithId(SchedulerIstanceId)
                    // https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html#quartz-scheduler-instancename
                    // Unique per Cluster "Group"
                    .WithName(SchedulerGroupName)
                    // default max concurrency is 1
                    .UseDefaultThreadPool(x => x.MaxConcurrency = 1)
                    // this is the default 
                    // .WithMisfireThreshold(TimeSpan.FromSeconds(60))
                    .UseJobFactory<MicrosoftDependencyInjectionJobFactory>()
                    .UsePersistentStore(x =>
                    {
                        // force job data map values to be considered as strings
                        // prevents nasty surprises if object is accidentally serialized and then 
                        // serialization format breaks, defaults to false
                        x.UseProperties = true;
                        x.UseClustering();
                        // there are other SQL providers supported too 
                        // Another reason why the Schema User and Schema Name need to match.
                        // Oracle Connection strings do not allow you to set the Schema Name.
                        x.UseOracle($"Data Source=(conn string);
                        // this requires Quartz.Serialization.Json NuGet package
                        x.UseJsonSerializer();
                    })
                    .BuildScheduler();
```
2023-02-13 12:57:52,162 [1] ERROR CI.HealthCore.Program - Quartz: Failled to start Scheduler
Quartz.SchedulerConfigException: Unable to Instantiate JobFactory: Cannot instantiate type which has no empty constructor (Parameter 'MicrosoftDependencyInjectionJobFactory')
 ---> System.ArgumentException: Cannot instantiate type which has no empty constructor (Parameter 'MicrosoftDependencyInjectionJobFactory')
   at Quartz.ExceptionHelper.ThrowArgumentException(String message, String paramName)
   at Quartz.Util.ObjectUtils.InstantiateType[T](Type type)
   at Quartz.Impl.StdSchedulerFactory.Instantiate()
   --- End of inner exception stack trace ---
   at Quartz.Impl.StdSchedulerFactory.Instantiate()
   at Quartz.Impl.StdSchedulerFactory.GetScheduler(CancellationToken cancellationToken)
   at CI.HealthCore.Services.QuartzService.StartServiceComplexAsync(ILog logger, ILogger msLogger) in \QuartzService.cs:line 64 [See nested exception: System.ArgumentException: Cannot instantiate type which has no empty constructor (Parameter 'MicrosoftDependencyInjectionJobFactory')
   at Quartz.ExceptionHelper.ThrowArgumentException(String message, String paramName)
   at Quartz.Util.ObjectUtils.InstantiateType[T](Type type)
   at Quartz.Impl.StdSchedulerFactory.Instantiate()]

**Reply by lahma** [reply]
Yeah,  so set via property and new the instance before scheduler start.

**Comment by s3YwCf2ZbfJG4SHAfjQMAjtsf** [comment]
Thanks for your help, but my situation might be slightly unique because I might not be in Program.cs with builder/services/app available to me and MicrosoftDependencyInjectionJobFactory has a constructor expecting ServiceProvider and options.

Thanks again.

**Comment by s3YwCf2ZbfJG4SHAfjQMAjtsf** [comment]
I think my problem is that although I am doing below in Program.cs, when I do "BuildScheduler()" later, it doesn't know that I want DependencyInjectionJobFactory.  I stumbled on your 4.0 code and SetLoggerFactory, that might be the solution given how I am doing things.   

Thanks again for trying..  :)

```
                QuartzService.SetSchedulerInstanceId();
                builder.Services.AddQuartz(q =>
                {
                    q.UseMicrosoftDependencyInjectionJobFactory();
                    q.SchedulerId = QuartzService.SchedulerInstanceId;
                });
                builder.Services.AddTransient<MyJob>();

                // Quartz.Extensions.Hosting allows you to fire background service that handles scheduler lifecycle
                builder.Services.AddQuartzHostedService(options =>
                {
                    // when shutting down we want jobs to complete gracefully
                    options.WaitForJobsToComplete = true;
                });
```

![image](https://user-images.githubusercontent.com/784740/218730099-500b2230-97bc-4d8d-b22b-7ceeadb925af.png)
